### PR TITLE
object-literal-sort-keys - bugfix & new options

### PR DESCRIPTION
**Pull request by SteveByerly**
#### PR checklist

- [ ] Addresses an existing issue: #0000
- [x] New feature, bugfix, or enhancement
  - [x] Includes tests
- [x] Documentation update

#### Overview of change:
Fixed a bug for numeric property types - the code was not checking `ts.SyntaxKind.NumericLiteral` for nodes.

The current rule didn't attempt to check object literals defined on a single-line, and this wasn't explicitly called out in the rule's description. Made this as a new option to keep backwards-compatibility vs just enabling it by default.

~~Added the ability to check whether shorthand properties are evaluated in order.~~
Added the ability to choose how object keys are sorted by type.
I originally used a rule from tslint-consistent-codestyle to require shorthand-properties first. The latest major update to tslint started conflicting with that rule. I really like keeping shorthand first since I think it helps reduce errors. I originally thought there could be two rules, but they would conflict to a degree. There could probably still be a shorthand-first rule for those not using sort-keys.

#### CHANGELOG.md entry:
[new-rule-option] `object-literal-sort-keys`: Add `check-single-line` option
[new-rule-option] `object-literal-sort-keys`: Add option to provide order of key types. Possible values are `spread`, `shorthand`, and `longhand`
[bugfix] `object-literal-sort-keys`: Properly compare numeric property names
[enhancement] `object-literal-sort-keys`: Do not reset after encountering a spread attribute

### CODE REVIEW

**Comment by palantirtech** [comment]
Thanks for your interest in [palantir/tslint](https://github.com/palantir/tslint), @SteveByerly! Before we can accept your pull request, you need to sign our contributor license agreement - just visit https://cla.palantir.com/ and follow the instructions. Once you sign, I'll automatically update this pull request.

**Comment by adidahiya** [comment]
I'm not a fan of the "shorthand", "longhand", and "spread" options... what's the point of them? That kind of code style seems cumbersome to maintain and the rule option feels too opinionated to include as an implementation that's fully supported in core TSLint.

The point of the sorting lint check is to help reduce merge conflicts, and while putting shorthand/longhand first can also accomplish this goal, I think it adds unnecessary cognitive load in working with the code.

Also... I don't think we should be linting for the position of spreads within object literals because the semantics change based on the ordering and this has the potential for a lot of confusion

**Comment by SteveByerly** [comment]
Actually, I completely disregarded the fact that spread order matters and cannot be forced to one spot in the object literal. That also would be the reason to re-start alphabetical sorting when a spread is encountered. I would like to update the PR to address this and remove a lot of the complexity to provided a specific sort order.

The reason for the PR is that our team has found it much easier to work with shorthand properties grouped before regular properties. To us ([and others](https://github.com/ajafff/tslint-consistent-codestyle#object-shorthand-properties-first)) having these properties intermixed hurts readability and add cognitive load.

We'd be unable to use the alpha-sorting rule without accounting for the position of shorthand. We could add it as our own custom rule, I just think there are likely others in the same position.

```
props = {
  className: 'special-btn',
  color: 'primary',
  isDisabled,
  isError: false,
  isValid,
}
```

```
props = {
  isDisabled,
  isValid,
  className: 'special-btn',
  color: 'primary',
  isError: false,
}
```

**Comment by ajafff** [comment]
> I originally used a rule from tslint-consistent-codestyle to require shorthand-properties first. The latest major update to tslint started conflicting with that rule.

I'm glad you liked my rule. And sorry for breaking you.

I found it pretty useful to have all shorthand properties in one place. It helps a lot when quickly scanning through the source. Everyone who has worked in or with my team has adopted that style.
I found some other projects that use this rule, too.

I'm +1 for adding an option for shorthand properties to the rule. They should also be checked for alphabetical order (I think this PR does that already).

-------

Re: sorting of numeric keys

I intentionally left them untouched in my last refactor of the rule. I wasn't sure how to sort them.
* numeric sort (ascending)
  * how to handle string keys that are actually numeric values? should they be handled as numeric or string?
* alpha sort by value (1e4 -> 10000): `1e4 < 2e-1`
* alpha sort by user written format (1e4 = 1e4): `1e4 > 2e-1`

-------
Some other things to consider:

* need to put back the special handling for object spread as mentioned by @adidahiya 
* If you want to sort numeric keys, you should consider making them sortable like shorthand properties : `[ "shorthand", "numeric", "regular"]`

**Comment by SteveByerly** [comment]
@ajafff Thank you for the input. Also thanks for putting the codestyle rules together - we really like them.

Re: spread - totally spaced on the necessity of it's order. I like the addition of numeric as an option, though.

Re: numeric sorting - do you have an idea of what you think the best approach is?
I feel like the current implementation is pretty reasonable, as the various assignments that could be made aren't necessarily equal.

``` javascript
> obj = { 2e-1: 'a',  "2e-1": 'b' };
> Object {0.2: "a", 2e-1: "b"}
```

``` javascript
const numbers = {
    2e-1: true,
    "2e-1": true,
    1: true,
    ~ [The key '1' is not sorted alphabetically]
    100: true,
}
```

**Comment by SteveByerly** [comment]
@adidahiya @ajafff I think I tried to push too much into this to make it overly customizable. I reverted all the specific sorting options and simply left the `shorthand-first` and `check-single-line` options which fill the needs I originally had. I think it works well and could be modified in the future if anyone actually had the need.

Let me know if you want me to back-merge and/or rebase.

**Comment by SteveByerly** [comment]
@adidahiya @ajafff would you be willing to re-review?

**Comment by adidahiya** [comment]
@SteveByerly sorry for the delay; I'll take a look today.

**Comment by giladgray** [comment]
@SteveByerly  please update this branch so we can review it, or close it if not relevant.

**Comment by SteveByerly** [comment]
I was waiting since August for it to get reviewed. If it's still a desirable feature, i'll spend the time updating. otherwise, feel free to close.

thanks!

File: src/rules/objectLiteralSortKeysRule.ts [hunk 1]
```diff
@@ -15,74 +15,170 @@
...
+        rationale: Lint.Utils.dedent`
+            Keeping object literal properties alphabetized can be useful in ensuring
+            consistency and can help prevent merge conflicts. In addition, grouping
+            shorthand properties first can improve readability and help prevent errors.`,
+        optionsDescription: Lint.Utils.dedent`
+            Three arguments may be optionally provided:
+
+            * \`"${OPTION_CHECK_SINGLE_LINE}"\`: Check objects defined on a single line.
+            * \`"${OPTION_IGNORE_CASE}"\`: Ignore case when comparing keys.
+            * \`"${OPTION_SHORTHAND_FIRST}"\`: Enforces shorthand properties appear first.
```
Comment (giladgray) [inline src/rules/objectLiteralSortKeysRule.ts]: add an example? "shorthand properties appear first: `{ name, age: 45 }`"

File: src/rules/objectLiteralSortKeysRule.ts [hunk 2]
```diff
@@ -15,74 +15,170 @@
...
+        // Only check single-line object literals if explicitly asked
+        if (!checkSingleLine && isSameLine(sourceFile, node.properties.pos, node.end)) {
+            return false;
+        }
+
+        return true;
+    }
+
+    function shouldCheckSyntax(currentProperty: SortableProperty, lastProperty?: SortableProperty) {
+        // Only check syntax if option is set, we have a property to check ,and they changed
```
Comment (giladgray) [inline src/rules/objectLiteralSortKeysRule.ts]: `check[, ]and`

File: src/rules/objectLiteralSortKeysRule.ts [hunk 3]
```diff
@@ -15,74 +15,170 @@
...
+    function cb(node: ts.Node): void {
+        if (!canCheckNode(node)) {
+            return ts.forEachChild(node, cb);
+        }
+
+        let lastKey: string | undefined;  // to look back for spelling
+        let lastProperty: SortableProperty | undefined; // to look back for assignment
+
+        const { properties } = node as ts.ObjectLiteralExpression;
+        outer: for (const property of properties) {
```
Comment (giladgray) [inline src/rules/objectLiteralSortKeysRule.ts]: 😱 is there any way to avoid this `outer:` syntax? perhaps another helper function?

**Comment by giladgray** [comment]
@SteveByerly there are some great bug fixes in here so please continue, if you'd be so kind! but now you've got merge conflicts since this rule has been touched by others in the meantime, so check that out first. i'm happy to help.

**Comment by adidahiya** [comment]
closing as stale, sorry it took so long to review @SteveByerly. let us know if it's ok for someone to take take your code and make a new PR.

**Pull request closed without merging**